### PR TITLE
Add CSRF_COOKIE to prevent csrf when using JWT_AUTH_COOKIE

### PR DESCRIPTION
**Pull request by bmpenuelas**
To prevent Cross-Site Request Forgery when using JWT_AUTH_COOKIE, the `csrftoken` cookie will also be set when issuing the JWT authentication token.

### CODE REVIEW

**Comment by yurikulaz** [comment]
I think this changes are not enough for CSRF protection when JWT_AUTH_COOKIE option is enabled.  Your patch adds csrf cookie, but looks like it's value will be never checked, because django-rest-framework explicitly wraps every view in csrf_exempt().

https://github.com/encode/django-rest-framework/blob/master/rest_framework/views.py#L134 

I believe that BaseJSONWebTokenAuthentication authentication class should enforce csrf validation when needed. Like in rest_framework.authentication.SessionAuthentication, which explicitly call CSRFCheck?

**Comment by jheld** [comment]
@bmpenuelas Thanks for writing this patch! Is this PR a comprehensive fix (with using the csrf decorator)? Is there any reason why I shouldn't make a custom `.whl` for this before merge? Is there a timeline for merge?

**Comment by bmpenuelas** [comment]
Yes @jheld , using the csrf decorator this fix alone can provide the desired protection.

**Comment by paul110590** [comment]
Does this protect against login csrf?

**Comment by paul110590** [comment]
Hi @bmpenuelas , what I mean is the CSRF token is not set until the JWT is issued, i.e. _after_ the user has logged in. Therefore an attacker could log in via CSRF and the user would think they were logged in to their own account. The user would then unwittingly be performing actions on the attacker's account. Or is there some protection against this that I am missing?

**Comment by bmpenuelas** [comment]
Protection is generally geared toward avoiding forging user requests. That protection is provided by the Anti-CSRF token and the JWT token working together. 

In the particular case you mention, at no time is the user session compromised. It's not a threat like user-impersonating CSRF, but it's true it can have privacy implications. I don't want to go off-topic here since this isn't related to the pull-request itself, but CSRFToken can protect your application from this situation too. For example, you can issue an Anti-CSRF token for the login process, and store it as both: a hidden form value, and a cookie. To validate the login request, both values must match, and since an attacker can't read them both, the login will fail.

**Comment by paul110590** [comment]
It's a different type of csrf to the more common session riding, but it's still a security flaw we need to protect against.

This is where I think it has relevancy to the pull request: To protect against login-csrf we need to decorate our login view with csrf_protect. We then need to create a view that will set the csrftoken cookie, so that we can fetch it prior to login and our login request is not blocked. This being the case, we no longer need to set the csrftoken cookie when we issue the JWT, because it is already set.

Basically, if we follow the approach of protecting ourselves from login-csrf we will also automatically be protecting ourselves from 'normal' csrf (provided we guard all unsafe views with csrf_protect). Whereas if we follow the approach of protecting from 'normal' csrf, we are still leaving ourselves open to login-csrf.

**Comment by bmpenuelas** [comment]
Yes, this PR already does that. CSRF tokens are issued for every request, not only after login, so you can use them for login, or to protect other requests, the same principles apply.

The problem before this PR was that when you used the JWT token for authentication, the CSRF token was ignored.

File: rest_framework_jwt/views.py [hunk 1]
```diff
@@ -3,6 +3,8 @@
 from rest_framework.response import Response
 from datetime import datetime
 
+from django.middleware import csrf
```
Comment (jayvdb) [inline rest_framework_jwt/views.py]: just a consideration as I saw this referred to at https://github.com/Styria-Digital/django-rest-framework-jwt/issues/4 , maybe it is appropriate to only import this if `api_settings.CSRF_COOKIE` is enabled, e.g. as a delayed import within the `if` below.  

Then any site not using csrf middleware doesnt get hit with an extra module being import that isnt used.  Are there many sites doing that?

Also pep8 tools would complain bitterly about that.

**Comment by pc-trent** [comment]
Any ideas when this PR is going to be merged into the package? Does it need work, can I help in any way getting this out the door?

**Comment by jayvdb** [comment]
@pc-trent offer sacrifices to @gcollazo and @jpadilla . c.f. https://github.com/GetBlimp/django-rest-framework-jwt/issues/449

**Comment by pc-trent** [comment]
> @pc-trent offer sacrifices to @gcollazo and @jpadilla . c.f. #449

I'll take that as a "this project is dead"

**Comment by jheld** [comment]
We should hope not. Maybe we can ask for maintainer privileges? Including PyPI. We've been waiting for this fix for a year.
